JiveAuthenticatingHTTPProtocolDemo/JiveAuthenticatingHTTPProtocolDemo/ViewController.m

```diff
@@ -9,142 +9,269 @@
+- (JAHPDidCancelAuthenticationChallengeHandler)authenticatingHTTPProtocol:(JAHPAuthenticatingHTTPProtocol *)authenticatingHTTPProtocol
+                                        didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge
+{
     self.authenticatingHTTPProtocol = authenticatingHTTPProtocol;
     
-    self.authAlertView = [[UIAlertView alloc] initWithTitle:@"JAHPDemo"
-                                                    message:@"Enter 'foo' for the username and 'bar' for the password"
-                                                   delegate:self
-                                          cancelButtonTitle:@"Cancel"
-                                          otherButtonTitles:
-                          @"OK",
-                          nil];
+    //=== result handler
+    //
+    __weak ViewController* weakSelf = self;
+    JAHPDidCancelAuthenticationChallengeHandler result =
+    ^void(
+          JAHPAuthenticatingHTTPProtocol* authenticatingHTTPProtocol,
+          NSURLAuthenticationChallenge*   challenge)
+    {
+        [weakSelf handleAuthChallenge: challenge
+                          forProtocol: authenticatingHTTPProtocol];
+    };
+
+    
+    //=== try using the stored credentials
+    //    to avoid entering them all over again
+    //
+    if (nil != self.userInput)
+    {
+        [self passCredentialsInputToConnection];
+        return result;
+    }
+    
+    
+    //=== show alert for the first time
+    //    if we have no credentials yet
+    //
+    self.authAlertView =
+    [[UIAlertView alloc] initWithTitle: @"JAHPDemo"
+                               message: @"Enter 'foo' for the username and 'bar' for the password"
+                              delegate: self
+                     cancelButtonTitle: @"Cancel"
+                     otherButtonTitles: @"OK", nil];
+    
     self.authAlertView.alertViewStyle = UIAlertViewStyleLoginAndPasswordInput;
     [self.authAlertView show];
     
-    return ^(JAHPAuthenticatingHTTPProtocol *authenticatingHTTPProtocol, NSURLAuthenticationChallenge *challenge) {
-        [self.authAlertView dismissWithClickedButtonIndex:self.authAlertView.cancelButtonIndex
-                                                 animated:YES];
-        self.authAlertView = nil;
-        self.authenticatingHTTPProtocol = nil;
-        
-        [[[UIAlertView alloc] initWithTitle:@"JAHPDemo"
-                                    message:@"The URL Loading System cancelled authentication"
-                                   delegate:nil
-                          cancelButtonTitle:@"OK"
-                          otherButtonTitles:nil] show];
-    };
+    return result;
+}
+
+
+-(void)handleAuthChallenge:(NSURLAuthenticationChallenge*)challenge
+               forProtocol:(JAHPAuthenticatingHTTPProtocol*)authenticatingHTTPProtocol
+{
+    // TODO: maybe fix leaking `self`
```

Comment:
Where is the leaking `self`?

Reply:
Same thing. Old irrelevant comment I've written before the refactoring happened. Will delete that.